lib/active_merchant/billing/gateways/paypal/paypal_common_api.rb

```diff
@@ -586,11 +586,11 @@ def add_payment_details(xml, money, currency_code, options = {})
           xml.tag! 'n2:OrderTotal', localized_amount(money, currency_code), 'currencyID' => currency_code
 
           # All of the values must be included together and add up to the order total
-          if [:subtotal, :shipping, :handling, :tax].all?{ |o| options.has_key?(o) }
-            xml.tag! 'n2:ItemTotal', localized_amount(options[:subtotal], currency_code), 'currencyID' => currency_code
-            xml.tag! 'n2:ShippingTotal', localized_amount(options[:shipping], currency_code),'currencyID' => currency_code
-            xml.tag! 'n2:HandlingTotal', localized_amount(options[:handling], currency_code),'currencyID' => currency_code
-            xml.tag! 'n2:TaxTotal', localized_amount(options[:tax], currency_code), 'currencyID' => currency_code
+          if [:subtotal, :shipping, :handling, :tax].all?{ |o| options[o].present?}
+            xml.tag! 'n2:ItemTotal', localized_amount(options[:subtotal].to_i, currency_code), 'currencyID' => currency_code
+            xml.tag! 'n2:ShippingTotal', localized_amount(options[:shipping].to_i, currency_code),'currencyID' => currency_code
+            xml.tag! 'n2:HandlingTotal', localized_amount(options[:handling].to_i, currency_code),'currencyID' => currency_code
+            xml.tag! 'n2:TaxTotal', localized_amount(options[:tax].to_i, currency_code), 'currencyID' => currency_code
```

Comment:
is there ever a chance that these values will need to be more precise than an integer? like for say, a three decimal currency?

Reply:
Good question. It looks like Paypal currently doesn't support three-decimal currencies (BHD, IQD, JOD, KWD, LYD, OMR, TND): https://developer.paypal.com/docs/api/reference/currency-codes/.  Also, the `localized_amount` method is expecting an integer here according to this line: https://github.com/activemerchant/active_merchant/blob/master/lib/active_merchant/billing/gateway.rb#L261. It'll convert those values back to a decimal before being sent to the Paypal gateway, so I think we're okay here in case they ever do decide to start supporting three-decimal currencies.